api/storage/client.go

```diff
@@ -126,50 +150,248 @@ func UploadArtifactWithOpts(ctx context.Context, c Client, namespace, path strin
+
+			r, ai, err := ResolveArtifactWithOpts(ctx, cli, newest.GetNamespace(), newest.GetPath(), ResolveArtifactOpts{})
+			if err != nil {
+				return nil, CacheInfo{}, err
+			}
+
+			return r, CacheInfo{CachedAt: ai.CreatedAt}, nil
+		} else {
+			logf("Content cached at %v is too old, loading from source...\n", newest.GetCreatedAt().AsTime())
+			// Fallthrough
+		}
+	} else {
+		logf("Artifact not found in cache; loading from source...\n")
+		// Fallthough
+	}
+
+	cachePath := cacheArtifactPath(time.Now(), parsedURL)
+	labels := map[string]string{cacheSourceURLLabel: sourceURL}
+
+	req, err := http.NewRequestWithContext(ctx, "GET", sourceURL, nil)
+	if err != nil {
+		return nil, CacheInfo{}, fmt.Errorf("failed to prepare request: %w", err)
+	}
+
+	resp, err := http.DefaultClient.Do(req)
+	if err != nil {
+		return nil, CacheInfo{}, CacheSourceError{fmt.Errorf("failed to send request: %w", err), 0}
+	}
+	defer resp.Body.Close()
+
+	if resp.StatusCode != http.StatusOK {
+		return nil, CacheInfo{}, CacheSourceError{fmt.Errorf("remote server returned status code %d", resp.StatusCode), resp.StatusCode}
+	}
+
+	w, err := os.CreateTemp("", "cache-")
+	if err != nil {
+		return nil, CacheInfo{}, fmt.Errorf("failed to open temp file: %w", err)
+	}
+
+	if err := os.Remove(w.Name()); err != nil {
+		return nil, CacheInfo{}, fmt.Errorf("failed to unlink temp file: %w", err)
+	}
+
+	success := false
+	defer func() {
+		if !success {
+			w.Close()
+		}
+	}()
+
+	// Uploading the response will write it to the file as a side-effect.
+	// Read errors will be tagged to allow the client to retry if needed.
+	teeR := io.TeeReader(wrapErrorsReader{resp.Body}, w)
+
+	var length int64
+	if resp.ContentLength >= 0 {
+		length = resp.ContentLength
+	}
+
+	ai, err := UploadArtifactWithOpts(ctx, cli, cacheArtifactNamespace, cachePath, teeR, UploadOpts{
+		Labels: labels,
+		Length: length,
+	})
+	if err != nil {
+		return nil, CacheInfo{}, fmt.Errorf("failed to cache artifact: %w", err)
+	}
+
+	if opts.ExpectedSHA256 != "" && opts.ExpectedSHA256 != ai.DigestSHA256 {
+		return nil, CacheInfo{}, status.Errorf(codes.FailedPrecondition, "artifact downloaded from source doesn't match expected digest: got %s, want %s", ai.DigestSHA256, opts.ExpectedSHA256)
+	}
+
+	if _, err := w.Seek(0, io.SeekStart); err != nil {
+		return nil, CacheInfo{}, fmt.Errorf("failed to rewind the temp file: %w", err)
+	}
+
+	success = true
+	return w, CacheInfo{CachedAt: ai.CreatedAt}, nil
+}
+
+type CacheSourceError struct {
+	Err            error
+	HTTPStatusCode int
+}
+
+func (err CacheSourceError) Error() string {
+	return err.Err.Error()
+}
+
+func (err CacheSourceError) Unwrap() error {
+	return err.Err
+}
+
+type wrapErrorsReader struct {
+	io.Reader
+}
+
+func (r wrapErrorsReader) Read(p []byte) (n int, err error) {
+	n, err = r.Reader.Read(p)
+	if errors.Is(err, io.EOF) {
+		// io.Copy checks for io.EOF equality
+		return n, err
+	}
+	if err != nil {
+		return n, CacheSourceError{err, 0}
```

Comment:
I would just return `err` here, not the 0 code.

Reply:
We are wrapping all the errors that come from interaction with the source server. It could drop while sending the request, err out with an HTTP status or drop while sending the body. The user is free to retry or not retry.